csharp/src/Ice/LoggerAdmin.cs

```diff
@@ -4,27 +4,99 @@
+
+        internal bool IsAccepted(LogMessage logMessage)
+        {
+            if (_messageTypes.Count == 0 || _messageTypes.Contains(logMessage.Type))
+            {
+                if (logMessage.Type != LogMessageType.TraceMessage || _traceCategories.Count == 0 ||
+                   _traceCategories.Contains(logMessage.TraceCategory))
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        internal void Queue(Func<IRemoteLoggerPrx, ValueTask> func, ILogger logger, string op)
+        {
+            lock (this)
+            {
+                _task = PerformAsync(func, logger, op);
+            }
+        }
+
+        private async ValueTask PerformAsync(Func<IRemoteLoggerPrx, ValueTask> func, ILogger logger, string op)
+        {
+            try
+            {
+                // Wait for the previous invocation to be sent.
+                await _task;
+
+                // Now send the given invocation.
+                await func(_remoteLoggerPrx);
+            }
+            catch (Exception ex)
+            {
+                _loggerAdmin.DeadRemoteLogger(_remoteLoggerPrx, logger, ex, op);
+                throw;
+            }
+        }
+    }
+
     internal sealed class LoggerAdmin : ILoggerAdmin
     {
-        public void
-        AttachRemoteLogger(IRemoteLoggerPrx? prx, LogMessageType[] messageTypes, string[] categories,
-                           int messageMax, Current current)
+        private const string TraceCategory = "Admin.Logger";
+        private bool _destroyed = false;
+        private int _logCount = 0; // non-trace messages
+        private readonly LoggerAdminLogger _logger;
+        private readonly int _maxLogCount;
+        private readonly int _maxTraceCount;
+        private LinkedListNode<LogMessage>? _oldestTrace = null;
+        private LinkedListNode<LogMessage>? _oldestLog = null;
+        private readonly LinkedList<LogMessage> _queue = new LinkedList<LogMessage>();
+        private readonly Dictionary<Identity, RemoteLoggerQueue> _remoteLoggerMap
+            = new Dictionary<Identity, RemoteLoggerQueue>();
+        private Communicator? _sendLogCommunicator = null;
+        private int _traceCount = 0;
+        private readonly int _traceLevel;
+
+        public void AttachRemoteLogger(IRemoteLoggerPrx? prx, LogMessageType[] types, string[] categories,
+            int messageMax, Current current)
         {
             if (prx == null)
             {
                 return; // can't send this null RemoteLogger anything!
             }
 
-            IRemoteLoggerPrx remoteLogger = prx.Clone(oneway: false);
+            var messageTypes = new HashSet<LogMessageType>(types);
+            var traceCategories = new HashSet<string>(categories);
 
-            var filters = new Filters(messageTypes, categories);
             LinkedList<LogMessage>? initLogMessages = null;
-
+            RemoteLoggerQueue remoteLogger;
```

Comment:
That's a poor choice of variable name.  If you want to keep queue, it should be remoteLoggerQueue. My preference would be `LogForwarder logForwarder`.

Reply:
Ok I also renamed _remoteLoggerMap to _logForwardMap and RemoteRemoteLogger to RemoveLogForwarder.